abstract_domains/numerical/octagon.py

```diff
@@ -0,0 +1,578 @@
+                self._var_to_index[var1] + _index_shift(sign1), self._var_to_index[var2] + _index_shift(sign2)
+            ]
+        else:
+            raise ValueError("Index into octagon has invalid format.")
+
+    def __setitem__(self, index_tuple: Tuple[Sign, VariableIdentifier, Sign, VariableIdentifier], value):
+        if len(index_tuple) == 4:
+            sign1, var1, sign2, var2 = index_tuple
+            i, j = self._var_to_index[var1] + _index_shift(sign1), self._var_to_index[var2] + _index_shift(sign2)
+            if i != j:
+                self.dbm[i, j] = value
+        else:
+            raise ValueError("Index into octagon has invalid format.")
+
+    def binary_constraints_indices(self, sign1: Sign = None, var1: VariableIdentifier = None,
+                                   sign2: Sign = None, var2: VariableIdentifier = None):
+        signs1 = [sign1] if sign1 else [PLUS, MINUS]
+        signs2 = [sign2] if sign2 else [PLUS, MINUS]
+        vars1 = [var1] if var1 else self.variables
+        vars2 = [var2] if var2 else self.variables
+        for var1 in vars1:
+            for var2 in vars2:
+                if var1 == var2:
+                    # do not yield diagonal and upper right triangular matrix
+                    continue
+                for sign1 in signs1:
+                    for sign2 in signs2:
+                        yield (sign1, var1, sign2, var2)
+
+    def __repr__(self):
+        if self.is_bottom():
+            return "BOTTOM"
+        elif self.is_top():
+            return "TOP"
+        else:
+            res = []
+            # represent unary constraints first
+            for var in self.variables:
+                lower = - self[PLUS, var, MINUS, var] // 2
+                upper = self[MINUS, var, PLUS, var] // 2
+                if lower < inf and upper < inf:
+                    res.append(f"{lower}≤{var.name}≤{upper}")
+                elif lower < inf:
+                    res.append(f"{lower}≤{var.name}")
+                elif upper < inf:
+                    res.append(f"{var.name}≤{upper}")
+            # represent binary constraints second, do not repeat identical inequalities
+            for i, var1 in enumerate(self.variables):
+                for j, var2 in enumerate(self.variables):
+                    if i > j:
+                        c = self[MINUS, var1, PLUS, var2]
+                        if c < inf:
+                            res.append(f"{var1.name}+{var2.name}≤{c}")
+                        c = self[MINUS, var1, MINUS, var2]
+                        if c < inf:
+                            res.append(f"{var1.name}-{var2.name}≤{c}")
+                        c = self[PLUS, var1, PLUS, var2]
+                        if c < inf:
+                            res.append(f"-{var1.name}+{var2.name}≤{c}")
+                        c = self[PLUS, var1, MINUS, var2]
+                        if c < inf:
+                            res.append(f"-{var1.name}-{var2.name}≤{c}")
```

Comment:
I do not really understand how this business with `PLUS` and `MINUS` works. Maybe comments would help.